File: app/src/main/java/com/example/spring/RecordUsageService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Record metered usage for a subscription feature during a billing period.
 */
@Service
public class RecordUsageService {

    private final EntitlementsRepository entitlementsRepository;
    private final ApplicationEventPublisher springEventPublisher;

    public RecordUsageService(EntitlementsRepository entitlementsRepository, ApplicationEventPublisher springEventPublisher) {
        this.entitlementsRepository = entitlementsRepository;
        this.springEventPublisher = springEventPublisher;
    }

    @Transactional
    public UsageMeter handle(RecordUsageRequest request) {
        // === GEN_REGION_START: service-body ===
        // spec-driven body for RecordUsage.
        // intent: Record metered usage for a subscription feature during a billing period.
        // requires: EntitlementsRepository
        // emits events: UsageRecorded
        // return: UsageMeter (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveUsageMeter.
        UsageMeter value = new UsageMeter(request.tenantId(), request.usageId(), request.subscriptionId(), request.accountId(), request.featureKey(), request.quantity(), request.periodStart(), request.periodEnd(), request.recordedAt());
        UsageMeter saved = this.entitlementsRepository.saveUsageMeter(value);
        this.springEventPublisher.publishEvent(new UsageRecorded(saved.accountId(), saved.featureKey(), saved.quantity(), saved.recordedAt(), saved.subscriptionId(), saved.tenantId(), saved.usageId()));
        return saved;
        // === GEN_REGION_END: service-body ===
    }
}

File: app-python/entitlements_app/services/record_usage_service.py
"""Record metered usage for a subscription feature during a billing period."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.record_usage_request import RecordUsageRequest
from entitlements_app.entities import UsageMeter
from entitlements_app.ports import EntitlementsRepository


class RecordUsageService:
    """Service handling Command `RecordUsage`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: RecordUsageRequest) -> UsageMeter:
        # spec-driven body for Command RecordUsage.
        # intent: Record metered usage for a subscription feature during a billing period.
        # requires: EntitlementsRepository
        # emits events: UsageRecorded
        # return: UsageMeter (entity); construct from request inputs and persist via EntitlementsRepository.saveUsageMeter.
        value = UsageMeter(tenantId=request.tenantId, usageId=request.usageId, subscriptionId=request.subscriptionId, accountId=request.accountId, featureKey=request.featureKey, quantity=request.quantity, periodStart=request.periodStart, periodEnd=request.periodEnd, recordedAt=request.recordedAt)
        return self.entitlements_repository.saveUsageMeter(value)
